### ISSUE TITLE
Future for unique tokens in DOMTokenList

### ISSUE THREAD

**Issue by ArkadiuszMichalski** [whatwg/dom#201]
Per DOM tokens in DOMTokenList should be unique, which is provided by [ordered set parser](https://dom.spec.whatwg.org/#concept-ordered-set-parser). But after long time already beyond this is not implemented by any engine, or mabye I wrong so please correct me (Edge, Webkit or Servo do this?).
Some bugs from Mozilla:
https://bugzilla.mozilla.org/show_bug.cgi?id=869788
https://bugzilla.mozilla.org/show_bug.cgi?id=910121

Repeated tokens cause that other definition around DOMTokenList are not valid, for example invoking `add()` should run the "update steps" (what means set associated attr to new value taken from ordered set serializer for tokens, so this attr's value should have unique tokens too). But browsers don't do this and we get repeated tokens. In the other site for `remove()` we get better behaviour (but not in all engine).

``` js
<script>

    var el = document.createElement("div");

    el.className = " a a b ";
    console.log(el.className.length); // 7
    console.log(el.classList.length); // 3
    console.log(Array.prototype.slice.call(el.classList).join(" ")); // "a a b"

    el.classList.add("a"); // should remove duplicate and white space in associated attr value

    console.log(el.className.length); // 7
    console.log(el.classList.length); // 3
    console.log(Array.prototype.slice.call(el.classList).join(" ")); // "a a b"

    el.classList.remove("a");

    console.log(el.className.length); // 1 (Gecko and Presto return 2, we have "b ")
    console.log(el.classList.length); // 1
    console.log(Array.prototype.slice.call(el.classList).join(" ")); // "b"

</script>
```

Unique tokens in DOMTokenList still are considered as something attainable in the near future?

**Comment by mgol** [comment]
FWIW, Safari Technical Preview has fixed all those bugs and is now following the spec.

**Comment by mgol** [comment]
This is what I get in Safari Tech Preview for this test case:
```js
var el = document.createElement("div");

el.className = " a a b ";
console.log(el.className.length); // 7
console.log(el.classList.length); // 2
console.log(Array.prototype.slice.call(el.classList).join(" ")); // "a b"

el.classList.add("a"); // should remove duplicate and white space in associated attr value

console.log(el.className.length); // 3
console.log(el.classList.length); // 2
console.log(Array.prototype.slice.call(el.classList).join(" ")); // "a b"

el.classList.remove("a");

console.log(el.className.length); // 1
console.log(el.classList.length); // 1
console.log(Array.prototype.slice.call(el.classList).join(" ")); // "b"
```

So it seems to be serializing very aggressively.

**Comment by ArkadiuszMichalski** [comment]
Still something is wrong, `console.log(el.className.length); // 3` << what value is here `a a b`? Per DOM, after invoke `add()` should be `a b`.

**Comment by mgol** [comment]
> Still something is wrong, `console.log(el.className.length); // 3` << what value is here `a a b`? Per DOM, after invoke `add()` should be `a b`.

It's `a b`. There are 3 characters there, don't forget the space. :-)

**Comment by ArkadiuszMichalski** [comment]
Oh right, sorry, so it's correct:)

**Comment by cdumez** [comment]
Yes, WebKit trunk always drops duplicates and cleans up whitespace currently. It looks like I need to change so that we only do this on add/remove/toggle/update/replace?

**Comment by cvrebert** [comment]
> Have you access to Edge, what result we get on this engine?

Edge 13.10586:

```js
var el = document.createElement("div");

el.className = " a a b ";
console.log(el.className.length); // 7
console.log(el.classList.length); // 3
console.log(Array.prototype.slice.call(el.classList).join(" ")); // "a a b"

el.classList.add("a"); // should remove duplicate and white space in associated attr value

console.log(el.className.length); // 7
console.log(el.classList.length); // 3
console.log(Array.prototype.slice.call(el.classList).join(" ")); // "a a b"

el.classList.remove("a");

console.log(el.className.length); // 1
console.log(el.classList.length); // 1
console.log(Array.prototype.slice.call(el.classList).join(" ")); // "b"
```

**Comment by ArkadiuszMichalski** [comment]
@mgol Can you report what Safari Tech Preview product for this:

```
<script>
	var el = document.createElement("div");
	el.classList = " a a b ";
	console.log(el.className.length); // 7
	console.log(el.classList.length); // 3
	console.log(el.attributes[0].value); // " a a b "
	console.log(Array.prototype.slice.call(el.classList).join(" ")); // "a a b"
</script>
```

**Comment by mgol** [comment]
@ArkadiuszMichalski 
```js
var el = document.createElement("div");
el.classList = " a a b ";
console.log(el.className.length); // 3
console.log(el.classList.length); // 2
console.log(el.attributes[0].value); // "a b "
console.log(Array.prototype.slice.call(el.classList).join(" ")); // "a b"
```

**Comment by ArkadiuszMichalski** [comment]
So, per DOM, it's wrong: setting `classList` should [PutForwards=value] and directly set `class` content attribute to passing value " a a b ", but here I see "a b ". But at least tokens in set are correct.
https://dom.spec.whatwg.org/#dom-domtokenlist-value
This behaviour fianlly cames from merging DOMSettableTokenList into DOMTokenList https://github.com/whatwg/dom/issues/119

**Comment by mgol** [comment]
@ArkadiuszMichalski Please read my post on GitHub, I pasted some things wrong and edited it immediately, you see the old thing when you look at the mail.

**Comment by ArkadiuszMichalski** [comment]
I use GitHub but I had to reload the page, ok, so I removed this "last space" question.

**Comment by cdumez** [comment]
@ArkadiuszMichalski : Why is the behavior wrong?

el.classList = " a a b ";
should be equivalent to calling:
el.classList.value = "a a b ";

shouldn't it?

in which case, el.attributes[0].value should probably return "a b".

If there is a bug, I don't mind fixing it but this looks like expected behavior to me.

**Comment by ArkadiuszMichalski** [comment]
@cdumez 
Yes, `el.classList` on setting is equivalent `el.classList.value` on setting, and DOM definition:
"Setting the value attribute must set an attribute value for the associated element using associated attribute’s local name and the given value."
I thought that in https://github.com/whatwg/dom/issues/105 you are focusing only on stringifier for `DOMTokenList, but yes, this fix https://github.com/whatwg/dom/commit/2920fc15b9e894c45ff84c5d3bb77f7513ff50e5 change `DOMSettableTokenList.value` too (but only getting, setting was change earlier, here https://github.com/whatwg/dom/commit/e48ced42e68ce540c8d7c05d43868082f9d86f11). This change was done probably because that was like stable browsers work (which implemented `DOMSettableTokenList` for some content attribute, now replaced by `DOMTokenList').

Changing stringifier for `DOMTokenList` doesn't need to enforce changing `DOMTokenList.value' . They don't need have to the same behavior, it all depends on what is more expected by the developers 
and compatible with actual content. Setting asociated attr through `DOMTokenList` to orginal passing value also can be useful.

Recently a lot of changes  were made around sets and hard to follow them, especially when browsers still operate in their own way. The current definition doesn't look bad and really, if others don't follow the new changes, it makes no sense to modify anything in DOM once again.

**Comment by annevk** [comment]
Okay to mark this as a duplicate of #105?

**Comment by ArkadiuszMichalski** [comment]
@annevk If https://github.com/whatwg/dom/issues/105 also fix this case https://github.com/whatwg/dom/issues/201#issuecomment-208049965 then you can close. Due  https://github.com/whatwg/dom/issues/201#issuecomment-206301673 Safari Technology Preview make progress here, but still not correct per spec.. I still look implementation of this at least in one engine in Windows (even in dev version).

**Comment by annevk** [comment]
@ArkadiuszMichalski I'm folding this into #105, if I'm overlooking something let me know, but my plan is to file bugs against all browsers apart from WebKit and see how that shakes out. Also, going forward I'll make sure (and you should remind me) to include tests for normative changes so we don't end up in this situation again. It's clearly not great and apologies for that.
